### PR TITLE
[SYCL] Allow USM ptrs in exclusive and inclusive scans

### PR DESCRIPTION
**Pull request by Michoumichmich**
The group algorithms are currently accepting only multi_ptrs while the SYCL spec (4.17.4) states that data ranges can be described using pointers, iterators or instances of the multi_ptr class. The spec allows to introduce additional restrictions, but now that this implementation supports USM should we still keep enforcing the use of multi_ptr ? The [doc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_algorithms.asciidoc) has no reference to the use of multi_ptr.

### CODE REVIEW

**Comment by dm-vodopyanov** [comment]
> now that this implementation supports USM should we still keep enforcing the use of multi_ptr ? The doc has no reference to the use of multi_ptr.

@Pennycook , can you please comment?

**Comment by Pennycook** [comment]
> @Pennycook , can you please comment?

Functionality LGTM.  Our implementation of `joint_reduce` already seems to accept pointers that aren't `multi_ptr`s, so it makes sense to extend that to `exclusive_scan` and `inclusive_scan`.

File: sycl/include/CL/sycl/group_algorithm.hpp [hunk 1]
```diff
@@ -791,14 +794,15 @@ joint_inclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
                      const ptrdiff_t &divisor) -> ptrdiff_t {
     return ((v + divisor - 1) / divisor) * divisor;
   };
-  typename InPtr::element_type x;
-  typename OutPtr::element_type carry = init;
+  typename std::remove_const<typename detail::remove_pointer<InPtr>::type>::type
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/group_algorithm.hpp:797]: Why do we remove constantness here? I can be illegal.

**Review by vladimirlaz — COMMENTED** [review]
> > @Pennycook , can you please comment?
> 
> Functionality LGTM. Our implementation of `joint_reduce` already seems to accept pointers that aren't `multi_ptr`s, so it makes sense to extend that to `exclusive_scan` and `inclusive_scan`.

should we "deduce" if the raw pointer is in global or local AS (for future performance optimizations)? Or this functionality assumes glocab AS?

File: sycl/include/CL/sycl/group_algorithm.hpp (hunk 1 above)
Comment (Michoumichmich) [reply sycl/include/CL/sycl/group_algorithm.hpp:797]: I did that to allow using a `const` input pointer. These two lines are used to get the underlying type, but the const qualifier is never casted away

Comment (Michoumichmich) [reply sycl/include/CL/sycl/group_algorithm.hpp:797]: Without it I get `/home/michel/sycl_workspace/deploy/bin/../include/sycl/CL/sycl/group_algorithm.hpp:781:1: note: candidate template ignored: requirement 'detail::is_arithmetic<std::plus<void>>::value' was not satisfied [with Group = cl::sycl::ext::oneapi::sub_group, InPtr = const float *, OutPtr = float *, BinaryOperation = float, T = std::plus<>]
joint_inclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,`

Comment (Michoumichmich) [reply sycl/include/CL/sycl/group_algorithm.hpp:797]: @vladimirlaz if you want I can remove that, the point was indeed to keep const correctness

Comment (Michoumichmich) [reply sycl/include/CL/sycl/group_algorithm.hpp:797]: Or I can use the type `typename detail::remove_pointer<OutPtr>::type` for `x` which shouldn't be const and avoid the issue

**Comment by Pennycook** [comment]
> should we "deduce" if the raw pointer is in global or local AS (for future performance optimizations)? Or this functionality assumes glocab AS?

Good question.  I think we should explore this in a separate PR, because my hope is that this won't be necessary.

The algorithms can work with any address space that all work-items in the group can see.  (We don't currently have a group that represents a single work-item, but if we introduced that degenerate case, even `private` memory would make sense I suppose).  Unlike the sub-group loads/stores, we never have to pass the pointer directly to a SPIR-V built-in -- if you take a peek inside the implementation, `joint_reduce` is basically just a loop around `reduce_over_group`, with all the pointer dereferences happening in the headers.

So my expectation is:
- If the user passes a `multi_ptr`, the compiler will propagate the address space to the point of dereference
- If the user passes a generic raw pointer and the compiler can deduce the address space at compile-time, it will propagate the address space to the point of dereference
- If the user passes a generic raw pointer and the compiler _can't_ deduce the address space at compile-time, it will generate a run-time check and hoist that check outside of the loop (because I _think_ it's safe to assume that incrementing a global pointer gives a global pointer)

I think we should only worry about adding our own run-time checks if the compiler is failing with the hoisting, and if we're certain that the compiler will then optimize away the run-time checks in the other cases.

File: sycl/include/CL/sycl/group_algorithm.hpp (hunk 1 above)
Comment (vladimirlaz) [reply sycl/include/CL/sycl/group_algorithm.hpp:797]: I got it. You can keep it as is.

Comment (Michoumichmich) [reply sycl/include/CL/sycl/group_algorithm.hpp:797]: > I got it. You can keep it as is.

Nice, you comment made me realise that I forgot some of these `remove_const`

**Comment by vladimirlaz** [comment]
> I think we should only worry about adding our own run-time checks if the compiler is failing with the hoisting, and if we're certain that the compiler will then optimize away the run-time checks in the other cases.

Thank you for the detailed explanation. It is clear now.

File: sycl/include/CL/sycl/group_algorithm.hpp (hunk 1 above)
Comment (Michoumichmich) [reply sycl/include/CL/sycl/group_algorithm.hpp:797]: > I got it. You can keep it as is.

Nice, thanks!

**Review by vladimirlaz — APPROVED** [review]

**Pull request merged**